File: java/src/main/java/neetcode/stack/p0155_minstack/MinStack.java
package neetcode.stack.p0155_minstack;

import java.util.Stack;

/**
 * A stack that supports push, pop, top, and getMin in O(1) time.
 *
 * Approach: two parallel stacks.
 * - {@code stack} holds all values in insertion order.
 * - {@code minStack} holds the running minimum at each stack level:
 *   {@code minStack.peek()} is always the minimum of all current elements.
 *
 * When pushing, the new minimum is min(val, current minimum).
 * When popping, both stacks are popped together — the level below already
 * has the correct minimum recorded for the remaining elements.
 *
 * Time:  O(1) for all operations
 * Space: O(n) — minStack mirrors the main stack in size
 */
public class MinStack {

    private final Stack<Integer> stack;
    private final Stack<Integer> minStack;

    public MinStack() {
        stack = new Stack<>();
        minStack = new Stack<>();
    }

    public void push(int val) {
        stack.push(val);
        if (minStack.isEmpty()) {
            minStack.push(val);
        } else {
            minStack.push(Math.min(val, minStack.peek()));
        }
    }

    public void pop() {
        stack.pop();
        minStack.pop();
    }

    public int top() {
        return stack.peek();
    }

    public int getMin() {
        return minStack.peek();
    }
}

File: python/src/neetcode/stack/p0155_minstack/min_stack.py
class MinStack:
    """
    Stack supporting push, pop, top and getMin in O(1).

    Approach: two parallel lists.
    - stack: holds all values in insertion order.
    - minStack: holds the running minimum at each level.
      minStack[-1] is always the minimum of all current elements.

    Time:  O(1) for all operations
    Space: O(n)
    """

    def __init__(self):
        self.stack = []
        self.minStack = []

    def push(self, val: int) -> None:
        self.stack.append(val)
        # new minimum is the smaller of val and current minimum
        minVal = min(val, self.minStack[-1] if self.minStack else val)
        self.minStack.append(minVal)

    def pop(self) -> None:
        self.stack.pop()
        self.minStack.pop()

    def top(self) -> int:
        return self.stack[-1]

    def getMin(self) -> int:
        return self.minStack[-1]
